File: src/mcp/auth/middleware.ts
/**
 * OAuth 2.1 Bearer Token Middleware for MCP
 *
 * Validates Bearer tokens on protected endpoints.
 */

import { createMiddleware } from "hono/factory";
import type { Context, Next } from "hono";
import { accessTokenStore } from "./stores";
import type { AccessToken } from "./types";

// Extend Hono context to include OAuth token info
declare module "hono" {
  interface ContextVariableMap {
    oauthToken?: AccessToken;
    oauthClientId?: string;
    oauthScope?: string;
    oauthUserId?: string;
  }
}

/**
 * Extract Bearer token from Authorization header
 */
function extractBearerToken(authHeader: string | undefined): string | null {
  if (!authHeader) return null;

  // Support "Bearer <token>" format
  if (authHeader.startsWith("Bearer ")) {
    return authHeader.slice(7);
  }

  // Also support just the token for convenience
  if (authHeader.startsWith("bearer ")) {
    return authHeader.slice(7);
  }

  return null;
}

/**
 * Bearer token authentication middleware
 *
 * Validates the Bearer token and adds token info to context.
 * Returns 401 if token is missing or invalid.
 */
export const bearerAuth = createMiddleware(async (c: Context, next: Next) => {
  const authHeader = c.req.header("authorization");
  const token = extractBearerToken(authHeader);

  if (!token) {
    return c.json(
      {
        error: "unauthorized",
        error_description: "Missing or invalid Authorization header. Use: Bearer <token>",
      },
      401,
      {
        "WWW-Authenticate": 'Bearer realm="mcp"',
      }
    );
  }

  // Validate token
  const accessToken = accessTokenStore.get(token);

  if (!accessToken) {
    return c.json(
      {
        error: "invalid_token",
        error_description: "Token is invalid or expired",
      },
      401,
      {
        "WWW-Authenticate": 'Bearer realm="mcp", error="invalid_token"',
      }
    );
  }

  // Add token info to context
  c.set("oauthToken", accessToken);
  c.set("oauthClientId", accessToken.client_id);
  c.set("oauthScope", accessToken.scope);
  c.set("oauthUserId", accessToken.user_id);

  await next();
});

/**
 * Optional bearer auth - validates token if present, but doesn't require it
 */
export const optionalBearerAuth = createMiddleware(async (c: Context, next: Next) => {
  const authHeader = c.req.header("authorization");
  const token = extractBearerToken(authHeader);

  if (token) {
    const accessToken = accessTokenStore.get(token);
    if (accessToken) {
      c.set("oauthToken", accessToken);
      c.set("oauthClientId", accessToken.client_id);
      c.set("oauthScope", accessToken.scope);
      c.set("oauthUserId", accessToken.user_id);
    }
  }

  await next();
});

/**
 * Scope validation middleware factory
 *
 * Use after bearerAuth to check if the token has required scope(s).
 */
export function requireScope(...requiredScopes: string[]) {
  return createMiddleware(async (c: Context, next: Next) => {
    const tokenScope = c.get("oauthScope");

    if (!tokenScope) {
      return c.json(
        {
          error: "insufficient_scope",
          error_description: "Token scope is missing",
        },
        403
      );
    }

    const grantedScopes = tokenScope.split(" ");

    // Check if any required scope is granted
    // Also check for parent scopes (mcp includes mcp:read and mcp:write)
    const hasScope = requiredScopes.some((required) => {
      if (grantedScopes.includes(required)) return true;
      // mcp scope grants all sub-scopes
      if (grantedScopes.includes("mcp") && required.startsWith("mcp:")) return true;
      return false;
    });

    if (!hasScope) {
      return c.json(
        {
          error: "insufficient_scope",
          error_description: `Required scope: ${requiredScopes.join(" or ")}`,
        },
        403,
        {
          "WWW-Authenticate": `Bearer realm="mcp", scope="${requiredScopes.join(" ")}"`,
        }
      );
    }

    await next();
  });
}

/**
 * Create conditional auth middleware that checks config
 */
export function conditionalAuth(enabled: boolean) {
  return createMiddleware(async (c: Context, next: Next) => {
    if (!enabled) {
      // Auth is disabled, allow all requests
      await next();
      return;
    }

    // Auth is enabled, require bearer token
    const authHeader = c.req.header("authorization");
    const token = extractBearerToken(authHeader);

    if (!token) {
      return c.json(
        {
          error: "unauthorized",
          error_description: "Authentication required. Use: Bearer <token>",
        },
        401,
        {
          "WWW-Authenticate": 'Bearer realm="mcp"',
        }
      );
    }

    const accessToken = accessTokenStore.get(token);

    if (!accessToken) {
      return c.json(
        {
          error: "invalid_token",
          error_description: "Token is invalid or expired",
        },
        401,
        {
          "WWW-Authenticate": 'Bearer realm="mcp", error="invalid_token"',
        }
      );
    }

    c.set("oauthToken", accessToken);
    c.set("oauthClientId", accessToken.client_id);
    c.set("oauthScope", accessToken.scope);
    c.set("oauthUserId", accessToken.user_id);

    await next();
  });
}

File: internal/oauth/middleware.go
package oauth

import (
	"context"
	"net/http"
	"strings"

	"github.com/huynle/brain-api/internal/storage"
)

// ---------------------------------------------------------------------------
// Context keys
// ---------------------------------------------------------------------------

type contextKey string

const (
	ctxOAuthClient contextKey = "oauthClientId"
	ctxOAuthScope  contextKey = "oauthScope"
)

// OAuthClientFromContext returns the OAuth client ID from the request context.
func OAuthClientFromContext(ctx context.Context) (string, bool) {
	v, ok := ctx.Value(ctxOAuthClient).(string)
	return v, ok
}

// OAuthScopeFromContext returns the OAuth scope string from the request context.
func OAuthScopeFromContext(ctx context.Context) (string, bool) {
	v, ok := ctx.Value(ctxOAuthScope).(string)
	return v, ok
}

// ---------------------------------------------------------------------------
// Token validator interface
// ---------------------------------------------------------------------------

// AccessTokenValidator validates OAuth access tokens against a backing store.
type AccessTokenValidator interface {
	GetAccessToken(ctx context.Context, tokenValue string) (*storage.OAuthAccessToken, error)
}

// ---------------------------------------------------------------------------
// Middleware
// ---------------------------------------------------------------------------

// BearerAuth returns middleware that requires a valid OAuth Bearer token.
// It extracts the token from the Authorization header, validates it against
// the token store, and sets oauthClientId and oauthScope in the request context.
// Returns 401 if the token is missing or invalid.
func BearerAuth(validator AccessTokenValidator) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			token := extractBearer(r.Header.Get("Authorization"))
			if token == "" {
				w.Header().Set("WWW-Authenticate", `Bearer realm="brain-api"`)
				writeOAuthError(w, http.StatusUnauthorized, "Unauthorized", "Missing authentication token")
				return
			}

			accessToken, err := validator.GetAccessToken(r.Context(), token)
			if err != nil {
				w.Header().Set("WWW-Authenticate", `Bearer realm="brain-api", error="invalid_token"`)
				writeOAuthError(w, http.StatusUnauthorized, "Unauthorized", "Invalid or expired access token")
				return
			}

			ctx := context.WithValue(r.Context(), ctxOAuthClient, accessToken.ClientID)
			ctx = context.WithValue(ctx, ctxOAuthScope, accessToken.Scope)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}

// OptionalBearerAuth returns middleware that validates an OAuth Bearer token
// if present, but passes through unauthenticated requests. If a token is
// provided and invalid, the request is rejected with 401.
func OptionalBearerAuth(validator AccessTokenValidator) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			token := extractBearer(r.Header.Get("Authorization"))
			if token == "" {
				// No token provided — pass through without auth context
				next.ServeHTTP(w, r)
				return
			}

			accessToken, err := validator.GetAccessToken(r.Context(), token)
			if err != nil {
				w.Header().Set("WWW-Authenticate", `Bearer realm="brain-api", error="invalid_token"`)
				writeOAuthError(w, http.StatusUnauthorized, "Unauthorized", "Invalid or expired access token")
				return
			}

			ctx := context.WithValue(r.Context(), ctxOAuthClient, accessToken.ClientID)
			ctx = context.WithValue(ctx, ctxOAuthScope, accessToken.Scope)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}

// RequireScope returns middleware that checks the authenticated token has all
// required scopes. The "mcp" scope grants access to all "mcp:*" sub-scopes.
// Returns 403 Forbidden if any required scope is missing.
// Must be used after BearerAuth or OptionalBearerAuth (when token is present).
func RequireScope(scopes ...string) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			grantedStr, ok := OAuthScopeFromContext(r.Context())
			if !ok {
				writeOAuthError(w, http.StatusForbidden, "Forbidden", "No OAuth scope in context")
				return
			}

			granted := parseScopeSet(grantedStr)

			for _, required := range scopes {
				if !hasScope(granted, required) {
					writeOAuthError(w, http.StatusForbidden, "Forbidden",
						"Insufficient scope: requires "+required)
					return
				}
			}

			next.ServeHTTP(w, r)
		})
	}
}

// ConditionalAuth returns BearerAuth when enabled is true, or a pass-through
// when disabled. Useful for toggling OAuth on MCP routes via configuration.
func ConditionalAuth(enabled bool, validator AccessTokenValidator) func(http.Handler) http.Handler {
	if enabled {
		return BearerAuth(validator)
	}
	return func(next http.Handler) http.Handler {
		return next
	}
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

// extractBearer extracts the token from a "Bearer <token>" Authorization header.
func extractBearer(authHeader string) string {
	if strings.HasPrefix(authHeader, "Bearer ") {
		return authHeader[7:]
	}
	if strings.HasPrefix(authHeader, "bearer ") {
		return authHeader[7:]
	}
	return ""
}

// parseScopeSet splits a space-delimited scope string into a set.
func parseScopeSet(scopeStr string) map[string]bool {
	set := make(map[string]bool)
	for _, s := range strings.Fields(scopeStr) {
		set[s] = true
	}
	return set
}

// hasScope checks if a required scope is satisfied by the granted set.
// The "mcp" scope implicitly grants access to all "mcp:*" sub-scopes.
func hasScope(granted map[string]bool, required string) bool {
	if granted[required] {
		return true
	}
	// "mcp" parent scope implies all "mcp:*" sub-scopes
	if strings.HasPrefix(required, "mcp:") && granted["mcp"] {
		return true
	}
	return false
}

// Note: writeOAuthError is defined in routes.go (same package).
// It writes {"error": errCode, "error_description": desc} per RFC 6749.
